lib/WP_Auth0_Options_Generic.php

```diff
@@ -1,69 +1,221 @@
+	}
+
+	/**
+	 * Get all the keys for constant-overridden settings.
+	 *
+	 * @return array
+	 */
+	public function get_all_constant_keys() {
+		return array_keys( $this->constant_opts );
+	}
+
+	/**
+	 * Get the option_name for the settings array.
+	 *
+	 * @return string
+	 */
 	public function get_options_name() {
 		return $this->_options_name;
 	}
 
+	/**
+	 * Return options from memory, database, defaults, or constants.
+	 *
+	 * @return array
+	 */
 	public function get_options() {
-		if ( empty( $this->_opt ) ) {
+		if ( empty( $this->_opts ) ) {
 			$options = get_option( $this->_options_name, array() );
+			if ( empty( $options ) || ! is_array( $options ) ) {
 
-			if ( ! is_array( $options ) ) {
+				// Brand new install, no saved options so get all defaults.
 				$options = $this->defaults();
-			}
+			} else {
 
-			$options = array_merge( $this->defaults(), $options );
+				// Make sure we have settings for everything we need.
+				$options = array_merge( $this->defaults(), $options );
+			}
 
-			$this->_opt = $options;
+			// Check for constant overrides and replace.
+			if ( ! empty( $this->constant_opts ) ) {
+				$options = array_replace_recursive( $options, $this->constant_opts );
+			}
+			$this->_opts = $options;
 		}
-		return $this->_opt;
+		return $this->_opts;
 	}
 
 	/**
-	 * Return a settings value
+	 * Return a filtered settings value or default.
 	 *
-	 * @param string $key - settings key
-	 * @param null   $default - what to return if a value is not set
+	 * @param string $key - Settings key to get.
+	 * @param mixed  $default - Default value to return if not found.
 	 *
 	 * @return mixed
+	 *
+	 * @link https://auth0.com/docs/cms/wordpress/extending#wp_auth0_get_option
 	 */
 	public function get( $key, $default = null ) {
 		$options = $this->get_options();
-		$value   = $default;
-
-		if ( isset( $options[ $key ] ) ) {
-			$value = $options[ $key ];
-		}
-
+		$value   = isset( $options[ $key ] ) ? $options[ $key ] : $default;
 		return apply_filters( 'wp_auth0_get_option', $value, $key );
 	}
 
+	/**
+	 * Update a setting if not already stored in a constant.
+	 * This method will fail silently if the option is already set in a constant.
+	 *
+	 * @param string $key - Option key name to update.
+	 * @param mixed  $value - Value to update with.
+	 * @param bool   $should_update - Flag to update DB options array with value stored in memory.
+	 *
+	 * @return bool
+	 */
 	public function set( $key, $value, $should_update = true ) {
-		$options         = $this->get_options();
+		$options = $this->get_options();
+
+		// Cannot set a setting that is being overridden by a constant.
+		if ( $this->has_constant_val( $key ) ) {
+			return false;
+		}
+
 		$options[ $key ] = $value;
-		$this->_opt      = $options;
+		$this->_opts     = $options;
 
-		if ( $should_update ) {
-			update_option( $this->_options_name, $options );
+		// No database update so process completed successfully.
+		if ( ! $should_update ) {
```

Comment:
is this update something you'll normally run after setting each key-value? Think of the case of setting a bunch of variables. I guess only the last call to this method would have this value as `true`. Right? Since I guess is not going to perform well doing so after each call. If so, would it make sense to remove this parameter from this method and just expose the `update_all` method for the user to call it directly on demand? You'll need to weight the most common use cases to make this decision.

Reply:
> is this update something you'll normally run after setting each key-value?

It's run most of the time, yes. 

> Since I guess is not going to perform well doing so after each call. 

Not excellent but we don't `set` much except in the initial setup process and, in that case, there are a number of HTTP requests that will be the main slow-down. The main time settings are saved is when the page is saved and, in that case, it's a single update. 

> If so, would it make sense to remove this parameter from this method

BC there so I'll punt that to when we refactor this whole system.